File: 236.java
// Leetcode 236 LOWEST COMMON ANCESTOR OF A BINARY TREE
// References:https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/solutions/65226/my-java-solution-which-is-easy-to-understand/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        if (root == null || root == p || root == q) return root;

        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);

        if(left != null && right != null) return root;

        return left != null ? left : right;

    }
}

File: 236.py
# Leetcode 236 LOWEST COMMON ANCESTOR OF A BINARY TREE
# References: https://www.youtube.com/watch?v=WO1tfq2sbsI

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        if not root:
            return None
        
        if root == q or root == q:
            return root
        
        l = lowestCommonAncestor(root.left, p, q)
        r = lowestCommonAncestor(root.right, p, q)

        if l and r:
            return root
        else:
            return l or r
    # Time: O(N)
    # Space: O(1) if not counting revursive stack frames, otherwise O(N)



        